haskell/private/dependencies.bzl

```diff
@@ -140,32 +140,55 @@ def gather_dep_info(ctx, deps):
       HaskellInfo: Unified information about all dependencies.
     """
 
+    package_ids = [
+        dep[HaskellLibraryInfo].package_id
+        for dep in deps
+        if HaskellLibraryInfo in dep
+    ]
+    package_databases = depset(transitive = [
+        dep[HaskellInfo].package_databases
+        for dep in deps
+        if HaskellInfo in dep
+    ])
+    static_libraries = depset(transitive = [
+        dep[HaskellInfo].static_libraries
+        for dep in deps
+        if HaskellInfo in dep
+    ])
+    dynamic_libraries = depset(transitive = [
+        dep[HaskellInfo].dynamic_libraries
+        for dep in deps
+        if HaskellInfo in dep
+    ])
+    interface_dirs = depset(transitive = [
+        dep[HaskellInfo].interface_dirs
+        for dep in deps
+        if HaskellInfo in dep
+    ])
+
     acc = HaskellInfo(
-        package_ids = set.empty(),
-        package_databases = set.empty(),
+        package_ids = package_ids,
+        package_databases = package_databases,
         version_macros = set.empty(),
-        static_libraries = [],
-        dynamic_libraries = set.empty(),
-        interface_dirs = set.empty(),
+        static_libraries = static_libraries,
+        dynamic_libraries = dynamic_libraries,
+        interface_dirs = interface_dirs,
         cc_dependencies = empty_HaskellCcInfo(),
         transitive_cc_dependencies = empty_HaskellCcInfo(),
     )
 
     for dep in deps:
         if HaskellInfo in dep:
             binfo = dep[HaskellInfo]
-            package_ids = acc.package_ids
             if HaskellLibraryInfo not in dep:
                 fail("Target {0} cannot depend on binary".format(ctx.attr.name))
-            if HaskellLibraryInfo in dep:
-                set.mutable_insert(package_ids, dep[HaskellLibraryInfo].package_id)
             acc = HaskellInfo(
-                package_ids = package_ids,
-                package_databases = set.mutable_union(acc.package_databases, binfo.package_databases),
+                package_ids = acc.package_ids,
+                package_databases = acc.package_databases,
```

Comment:
Why no union between `acc.package_databases` and `binfo.package_databases`?

Reply:
Because we already computed the set of all package databases a few lines earlier. Look at the other hunks as well.